File: python/src/train/models/esmm.py
from __future__ import annotations

"""Configurable ESMM: shared bottom + task towers + probability relations."""

from typing import TYPE_CHECKING

import torch
import torch.nn as nn

if TYPE_CHECKING:
    from ..core.config import PoolingMode

from ..core.model_output import ModelExecution, ModelOutput
from ..core.output_contract import NormalizedOutputContract
from ..layers.embedding import FeatureEmbeddings, FeatureTensorMap, FeatureTuple
from ..layers.mlp import Mlp
from ..layers.towers import Activation, MultiTaskConfig, TaskRelation, TaskTower, TowerConfig
from .output_head import OutputHead


def default_task_config(
    click_hidden_dims: list[int],
    cvr_hidden_dims: list[int],
    detail_hidden_dims: list[int],
    stock_hidden_dims: list[int],
    stay_hidden_dims: list[int],
) -> MultiTaskConfig:
    return MultiTaskConfig(
        towers=[
            TowerConfig("click", click_hidden_dims, 1, Activation.RELU),
            TowerConfig("cvr", cvr_hidden_dims, 1, Activation.RELU),
            TowerConfig("detail", detail_hidden_dims, 1, Activation.RELU),
            TowerConfig("stock", stock_hidden_dims, 1, Activation.RELU),
            TowerConfig("stay", stay_hidden_dims, 1, Activation.RELU),
        ],
        relations=[
            TaskRelation("ctcvr", ["click", "cvr"], "multiply"),
            TaskRelation("ctdetail", ["click", "detail"], "multiply"),
            TaskRelation("ctstock", ["click", "stock"], "multiply"),
            TaskRelation("ctstay", ["detail", "stay"], "multiply"),
        ],
    )


class ESMM(nn.Module):
    """Entire-space multi-task model with configurable towers and relations."""

    def __init__(
        self,
        features: list[FeatureTuple],
        shared_bottom_dims: list[int],
        click_hidden_dims: list[int],
        cvr_hidden_dims: list[int],
        detail_hidden_dims: list[int],
        stock_hidden_dims: list[int],
        stay_hidden_dims: list[int],
        task_config: MultiTaskConfig | None = None,
        pooling_map: dict[str, PoolingMode] | None = None,
        total_dim: int | None = None,
        output_contract: NormalizedOutputContract | None = None,
    ) -> None:
        super().__init__()
        self.embeddings = FeatureEmbeddings(features, pooling_map, total_dim=total_dim)
        if shared_bottom_dims:
            self.shared_bottom = Mlp(
                self.embeddings.total_dim,
                shared_bottom_dims[:-1],
                shared_bottom_dims[-1],
                Activation.RELU,
            )
            sd = shared_bottom_dims[-1]
        else:
            sd = self.embeddings.total_dim

        if output_contract is not None:
            self.output_contract = output_contract
            self.output_head = OutputHead(output_contract, {"shared": sd})
            self.task_names = [tower.name for tower in output_contract.towers]
            self.relation_names = list(output_contract.relation_order)
            return

        self.task_config = task_config or default_task_config(
            click_hidden_dims,
            cvr_hidden_dims,
            detail_hidden_dims,
            stock_hidden_dims,
            stay_hidden_dims,
        )
        self.task_names = [tower.name for tower in self.task_config.towers]
        self.relation_names = [relation.target for relation in self.task_config.relations]
        for tower in self.task_config.towers:
            setattr(self, f"{tower.name}_tower", TaskTower(tower, sd))

    def forward(self, x_inputs: FeatureTensorMap) -> ModelOutput:
        if hasattr(self, "output_head"):
            return self.forward_execution(x_inputs).outputs
        return self._forward_legacy(x_inputs)

    def forward_execution(self, x_inputs: FeatureTensorMap) -> ModelExecution:
        concat = self.embeddings(x_inputs)
        shared = self.shared_bottom(concat) if hasattr(self, "shared_bottom") else concat
        if hasattr(self, "output_head"):
            return self.output_head({"shared": shared})
        outputs = self._forward_legacy_from_shared(shared)
        return ModelExecution(nodes=outputs, outputs=outputs)

    def _forward_legacy(self, x_inputs: FeatureTensorMap) -> ModelOutput:
        concat = self.embeddings(x_inputs)
        shared = self.shared_bottom(concat) if hasattr(self, "shared_bottom") else concat
        return self._forward_legacy_from_shared(shared)

    def _forward_legacy_from_shared(self, shared: torch.Tensor) -> ModelOutput:
        outputs = ModelOutput()
        for name in self.task_names:
            tower = getattr(self, f"{name}_tower")
            outputs.insert(name, tower(shared), tower.output_kind)
        for relation in self.task_config.relations:
            outputs.insert_probability(relation.target, self._apply_relation(relation, outputs))
        return outputs

    @staticmethod
    def _apply_relation(relation: TaskRelation, outputs: ModelOutput) -> torch.Tensor:
        if not relation.sources:
            raise ValueError(f"Relation '{relation.target}' has no sources")
        probs = [
            _probability_for_relation(relation, outputs, source) for source in relation.sources
        ]
        if relation.op == "multiply":
            result = probs[0]
            for value in probs[1:]:
                result = result * value
            return result
        if relation.op == "add":
            result = probs[0]
            for value in probs[1:]:
                result = result + value
            return result
        if relation.op == "subtract":
            if len(probs) != 2:
                raise ValueError(f"Relation '{relation.target}' subtract requires 2 sources")
            return probs[0] - probs[1]
        if relation.op == "divide":
            if len(probs) != 2:
                raise ValueError(f"Relation '{relation.target}' divide requires 2 sources")
            return probs[0] / (probs[1] + 1e-8)
        raise ValueError(f"Unknown relation op: {relation.op}")


def _probability_for_relation(
    relation: TaskRelation, outputs: ModelOutput, source: str
) -> torch.Tensor:
    output = outputs.get(source)
    if output is None:
        raise ValueError(f"Relation '{relation.target}' source '{source}' is missing")
    if output.kind == "binary_logit":
        return torch.sigmoid(output.tensor)
    if output.kind == "probability":
        return output.tensor
    raise ValueError(
        f"Relation '{relation.target}' source '{source}' must be binary_logit or probability, "
        f"got {output.kind}"
    )

File: src/models/esmm.rs
//! Configurable ESMM: shared bottom + task towers + probability relations.
use super::{Model, ModelExecution, ModelOutput, OutputKind};
use crate::layers::embedding::{FeatureEmbeddings, FeatureSpec};
use crate::layers::mlp::Mlp;
use crate::layers::towers::{
    apply_relation, Activation, MultiTaskConfig, RelationOp, TaskRelation, TaskTower, TowerConfig,
};
use crate::models::output_contract::OutputContract;
use crate::models::output_head::OutputHead;
use candle_core::{Result, Tensor};
use candle_nn::VarBuilder;
use std::collections::HashMap;

/// 生成默认 5 任务 ESMM 配置 (click/cvr/detail/stock/stay)。
pub fn default_task_config(
    click_hidden_dims: &[usize],
    cvr_hidden_dims: &[usize],
    detail_hidden_dims: &[usize],
    stock_hidden_dims: &[usize],
    stay_hidden_dims: &[usize],
) -> MultiTaskConfig {
    MultiTaskConfig {
        towers: vec![
            TowerConfig {
                name: "click".into(),
                hidden_dims: click_hidden_dims.to_vec(),
                output_dim: 1,
                activation: Activation::Relu,
                output_kind: OutputKind::BinaryLogit,
            },
            TowerConfig {
                name: "cvr".into(),
                hidden_dims: cvr_hidden_dims.to_vec(),
                output_dim: 1,
                activation: Activation::Relu,
                output_kind: OutputKind::BinaryLogit,
            },
            TowerConfig {
                name: "detail".into(),
                hidden_dims: detail_hidden_dims.to_vec(),
                output_dim: 1,
                activation: Activation::Relu,
                output_kind: OutputKind::BinaryLogit,
            },
            TowerConfig {
                name: "stock".into(),
                hidden_dims: stock_hidden_dims.to_vec(),
                output_dim: 1,
                activation: Activation::Relu,
                output_kind: OutputKind::BinaryLogit,
            },
            TowerConfig {
                name: "stay".into(),
                hidden_dims: stay_hidden_dims.to_vec(),
                output_dim: 1,
                activation: Activation::Relu,
                output_kind: OutputKind::BinaryLogit,
            },
        ],
        relations: vec![
            TaskRelation {
                target: "ctcvr".into(),
                sources: vec!["click".into(), "cvr".into()],
                op: RelationOp::Multiply,
            },
            TaskRelation {
                target: "ctdetail".into(),
                sources: vec!["click".into(), "detail".into()],
                op: RelationOp::Multiply,
            },
            TaskRelation {
                target: "ctstock".into(),
                sources: vec!["click".into(), "stock".into()],
                op: RelationOp::Multiply,
            },
            TaskRelation {
                target: "ctstay".into(),
                sources: vec!["detail".into(), "stay".into()],
                op: RelationOp::Multiply,
            },
        ],
    }
}

/// Entire-space multi-task model with configurable towers and relations.
pub struct ESMM {
    embeddings: FeatureEmbeddings,
    shared_bottom: Option<Mlp>,
    towers: Vec<(String, TaskTower)>,
    relations: Vec<TaskRelation>,
}

impl ESMM {
    #[allow(clippy::too_many_arguments)]
    /// 使用默认任务配置构造 ESMM。
    pub fn new(
        vb: VarBuilder,
        features: &[FeatureSpec],
        shared_bottom_dims: &[usize],
        click_hidden_dims: &[usize],
        cvr_hidden_dims: &[usize],
        detail_hidden_dims: &[usize],
        stock_hidden_dims: &[usize],
        stay_hidden_dims: &[usize],
    ) -> Result<Self> {
        let task_config = default_task_config(
            click_hidden_dims,
            cvr_hidden_dims,
            detail_hidden_dims,
            stock_hidden_dims,
            stay_hidden_dims,
        );
        Self::with_task_config(vb, features, shared_bottom_dims, &task_config)
    }

    /// 使用自定义 MultiTaskConfig 构造 ESMM。
    pub fn with_task_config(
        vb: VarBuilder,
        features: &[FeatureSpec],
        shared_bottom_dims: &[usize],
        task_config: &MultiTaskConfig,
    ) -> Result<Self> {
        let embeddings = FeatureEmbeddings::new(vb.pp("embeddings"), features)?;
        let (shared_bottom, shared_output_dim) = if shared_bottom_dims.is_empty() {
            (None, embeddings.total_dim)
        } else {
            let output_dim = *shared_bottom_dims.last().unwrap();
            let mlp = Mlp::new(
                vb.pp("shared_bottom"),
                embeddings.total_dim,
                &shared_bottom_dims[..shared_bottom_dims.len() - 1],
                output_dim,
                Activation::Relu,
            )?;
            (Some(mlp), output_dim)
        };
        let mut towers = Vec::with_capacity(task_config.towers.len());
        for tower_config in &task_config.towers {
            towers.push((
                tower_config.name.clone(),
                TaskTower::new(
                    tower_config,
                    shared_output_dim,
                    vb.pp(format!("{}_tower", tower_config.name)),
                )?,
            ));
        }
        Ok(Self {
            embeddings,
            shared_bottom,
            towers,
            relations: task_config.relations.clone(),
        })
    }
}

impl Model for ESMM {
    fn forward(&self, x_inputs: &HashMap<String, Tensor>) -> Result<ModelOutput> {
        let concat = self.embeddings.forward(x_inputs)?;
        let shared_output = match &self.shared_bottom {
            Some(b) => b.forward(&concat)?,
            None => concat,
        };
        let mut outputs = ModelOutput::new();
        for (name, tower) in &self.towers {
            outputs.insert_binary_logit(name.clone(), tower.forward(&shared_output)?);
        }
        for relation in &self.relations {
            outputs
                .insert_probability(relation.target.clone(), apply_relation(relation, &outputs)?);
        }
        Ok(outputs)
    }
}

/// Standard ESMM backbone using the generic output contract execution path.
pub struct ContractEsmm {
    embeddings: FeatureEmbeddings,
    shared_bottom: Option<Mlp>,
    output_head: OutputHead,
}

impl ContractEsmm {
    /// Build a shared-bottom ESMM with contract-defined towers and relations.
    pub fn new(
        vb: VarBuilder,
        features: &[FeatureSpec],
        shared_bottom_dims: &[usize],
        contract: &OutputContract,
    ) -> Result<Self> {
        let embeddings = FeatureEmbeddings::new(vb.pp("embeddings"), features)?;
        let (shared_bottom, shared_output_dim) = if shared_bottom_dims.is_empty() {
            (None, embeddings.total_dim)
        } else {
            let output_dim = *shared_bottom_dims.last().ok_or_else(|| {
                candle_core::Error::Msg("shared_bottom_dims must not be empty".into())
            })?;
            let mlp = Mlp::new(
                vb.pp("shared_bottom"),
                embeddings.total_dim,
                &shared_bottom_dims[..shared_bottom_dims.len() - 1],
                output_dim,
                Activation::Relu,
            )?;
            (Some(mlp), output_dim)
        };
        let representation_dims = HashMap::from([("shared".to_string(), shared_output_dim)]);
        let output_head = OutputHead::new(contract, &representation_dims, vb.pp("output_head"))?;
        Ok(Self {
            embeddings,
            shared_bottom,
            output_head,
        })
    }

    fn representations(
        &self,
        x_inputs: &HashMap<String, Tensor>,
    ) -> Result<HashMap<String, Tensor>> {
        let concat = self.embeddings.forward(x_inputs)?;
        let shared = match &self.shared_bottom {
            Some(bottom) => bottom.forward(&concat)?,
            None => concat,
        };
        Ok(HashMap::from([("shared".to_string(), shared)]))
    }
}

impl Model for ContractEsmm {
    fn forward(&self, x_inputs: &HashMap<String, Tensor>) -> Result<ModelOutput> {
        Ok(self.forward_execution(x_inputs)?.outputs)
    }

    fn forward_execution(&self, x_inputs: &HashMap<String, Tensor>) -> Result<ModelExecution> {
        self.output_head.forward(&self.representations(x_inputs)?)
    }
}
